feat: add built-in caching & tile loading cancellation support

JaffaKetchup opened the pull request:
See https://docs.fleaflet.dev/layers/tile-layer/built-in-caching & https://github.com/fleaflet/flutter_map/blob/caching/lib/src/layer/tile_layer/tile_provider/network/caching/built_in/impl/native/README.md.

Adds 'simple' built-in caching to `NetworkTileProvider` for native platforms. On the web (or where the caching mechanism cannot be setup for whatever reason), the `NetworkTileProvider` falls back to its existing implementation. 

Unlike existing caching methods, this requires absolutely no additional code to use. Also, plugins can create their own caching providers to integrate with existing compatible tile providers.

Through testing, it seems to work reasonably well. The OSM tile server does seem to spit out some weird HTTP headers which seems to ask for unnecessary requests. I have opened a thread on the OSM Slack to try to figure out what's going on - no response, and it seems to work with other servers, so I'm opting to ignore it.

---

Additionally, the latest beta version of 'package:http' supports aborting in-flight requests. Therefore, this integrates with that, starting the deprecation of the CNTP plugin.

androidseb commented:
Hey @JaffaKetchup, since you asked for feedback about this, I'll try and add some thoughts here. Disclaimer: I read through the code at a high level, not in fine details, so apologies if I missed some things there.

I'm not sure why there is a concept of "loading" for the caching mechanism, is it because we need to read an index JSON file first? I find that option to not be great because the JSON will grow larger and larger as the user has more and more tiles, and as you pointed out, might get corrupted, and I'm hoping we don't need that, and I'm hoping we can build this in a way that doesn't have an initialization delay by using the file system directly.

Do we need to store anything more than those pieces of information for a given cached tile (I'm hoping not)?

* x
* y
* z
* timestamp

I've had to tackle a similar problem in my app before because I implemented offline maps which are expensive to process (it means generating a PNG image from vector data), so I cache offline built tiles so I don't have to reprocess them later. I have implemented a simple approach where I cache tiles based on file named `x_y_z.png`. I understand the use case is a little more complex here, but that approach might still work with minor adjustments. I saw you wanted to cache tiles based on their URL, but I'm hoping we don't need that and can just focus on x / y / z.

Here are some ideas I hope would help this work better:

* Android has a dedicated "cache" folder, meaning if you locate your cache files there, it can be deleted by the user without the need of implementing this in the app's code, so I would default the storage folder to there
* Get rid of the JSON index file completely, and use the file system as your cache
     * MapTileCachingManager would take in a cache folder path and be instantly initialized and available
     * When requested for a tile, check if the file named `x_y_z.png` exists, and
         * if the file is recent enough, return its content
         * if the file is too outdated (passed the cache expiry setting), return a cache miss response
     * In the background, run an indexing process to list out all the cached files inventory - this will delay the moment we can start enforcing cache max size, but not the moment we can start serving cached tiles
     * Once the cache index is built, we can run an outdated tiles + max size enforcement cleanup
     * When requested to store a tile
         * Store write the tile using the file named `x_y_z.png.tmp`
         * Then after the file write has fully completed, rename the file to `x_y_z.png` (this avoid having half-written files)
         * If the index is already built, delete oldest files if exceeding storage quota
* Can you get rid of isolates? Don't async file read / write methods do the same thing? It would make the code simpler if no negative impact on performance. Have we confirmed using isolates is actually positively impacting performance?

I might not have the full context here, so this might be a naive approach that doesn't work for some reason, but I hope this helps :-)

JaffaKetchup commented:
Hey @androidseb, thanks for your thoughts :)

> Do we need to store anything more than those pieces of information for a given cached tile

We need to store:
 * Some way to refer to the unique tile
 * The HTTP caching headers (or calculated result) associated with the tile (of which there are at least 2)
 * (Ideally) the time at which the tile metadata (above) was last updated in the cache

So unfortunately some form of additional storage is required for the caching headers

> I saw you wanted to cache tiles based on their URL, but I'm hoping we don't need that and can just focus on x / y / z.

This might not work well. If a user is using two different tile servers with the same coords, we need to differentiate that. We could create a parent directory for each URL, but I'm not sure I can see the value.

> Android has a dedicated "cache" folder, meaning if you locate your cache files there, it can be deleted by the user without the need of implementing this in the app's code, so I would default the storage folder to there

I'm using whatever directory is returned by path_provider's `getApplicationCacheDirectory`. I'm assuming it would be cleared when the user requests to clear the app cache.

> if the file is too outdated (passed the cache expiry setting), return a cache miss response

I did consider this kind of storage mechanism, but how would you store the expiry time? In another file, one per tile? I want this to cache based off HTTP headers primarily, to be compliant with the OSM requirements for example.

> In the background, run an indexing process to list out all the cached files inventory - this will delay the moment we can start enforcing cache max size, but not the moment we can start serving cached tiles

Would this be in an isolate? I've tried to minimise I/O on main thread (see below). But yeah, the method I've written for enforcing the max cache size would be much better if it didn't delay the caching - this is difficult to do performantly however with the current setup. Maybe a change in setup would better allow this.

> Can you get rid of isolates? Don't async file read / write methods do the same thing? It would make the code simpler if no negative impact on performance. Have we confirmed using isolates is actually positively impacting performance?

The idea is to eliminate some of the overheads. Although you've suggested removing the registry file, in its current form, opening, writing and closing the file for every single update would be quite costly. The async methods do unblock the main thread, but come with their own overheads. Using long-lived isolates avoids some of this. I'm not sure whether the tile file writer isolate makes a noticeable difference, but it seemed to have some good effect (not properly profiled), and theoretically it should avoid some overheads.

I'm definitely on board with looking into ways to remove the registry, but I just don't see a great way around it to store all of the required metadata - except maybe using two files per tile, but that introduces additional overheads as a file needs to be read before the initial caching decision at fetch time can be made (rather than just reading from memory). Let me know what you think!

androidseb commented:
Thanks, I think I understand this a little better now.

## About storage

> We need to store:
>
> * Some way to refer to the unique tile
> * The HTTP caching headers (or calculated result) associated with the tile (of which there are at least 2)
> * (Ideally) the time at which the tile metadata (above) was last updated in the cache

* Some way to refer to the unique tile: the file name should suffice hopefully?
* The HTTP caching headers (or calculated result) associated with the tile (of which there are at least 2): I guess you could split this into two files: `tile_x_y_z.png` and `tile_x_y_z.png.metadata`
* (Ideally) the time at which the tile metadata (above) was last updated in the cache: that would be the file creation / last modified time which we have with the file system entity, but could also be stored in the metadata file...

## About using files directly

> This might not work well. If a user is using two different tile servers with the same coords, we need to differentiate that. We could create a parent directory for each URL, but I'm not sure I can see the value.

I agree splitting URLs per directory will make things difficult, but we could maybe figure out a unique file naming function for that, even if it means including a file-name-friendly version of the URL in the file name, I don't think it would necessarily be an issue.

## About the android cache folder

> > Android has a dedicated "cache" folder, meaning if you locate your cache files there, it can be deleted by the user without the need of implementing this in the app's code, so I would default the storage folder to there
> 
> I'm using whatever directory is returned by path_provider's getApplicationCacheDirectory. I'm assuming it would be cleared when the user requests to clear the app cache.

Oh OK, good, I had missed that, my comment is not relevant then.

## About cache expiry based on headers

> > if the file is too outdated (passed the cache expiry setting), return a cache miss response
> 
> I did consider this kind of storage mechanism, but how would you store the expiry time? In another file, one per tile? I want this to cache based off HTTP headers primarily, to be compliant with the OSM requirements for example.

Yes, having a `.metadata` file associated with the main `.png` file would work for this hopefully.

## About I/O

> Although you've suggested removing the registry file, in its current form, opening, writing and closing the file for every single update would be quite costly.

What do you mean by costly? More costly than fetching the data from the web with an HTTP request over the network? The tiles are fairly small and light in volume (~50kb for a 256x256 tile) - I can't imagine fetching cached tile for the entire screen of a device would exceed 10MB of reading on the disk, and that would almost always be faster than re-fetching them from the network anyways, right?

> The async methods do unblock the main thread, but come with their own overheads. Using long-lived isolates avoids some of this. I'm not sure whether the tile file writer isolate makes a noticeable difference, but it seemed to have some good effect (not properly profiled), and theoretically it should avoid some overheads.

What kind of overhead are you thinking about? I'm not super knowledgeable about these things so I'd be curious to have insights about this. I have a hunch that isolates might have noticeable value in intensive I/O operations, but given the size and number of the tiles we're dealing with here, I feel like it will not make a difference for the user, especially if you compare it with network-based requests. It might be worth keeping them for later if needed, instead of preemptively making them part of the architecture, but it's just a thought.

> I'm definitely on board with looking into ways to remove the registry, but I just don't see a great way around it to store all of the required metadata - except maybe using two files per tile, but that introduces additional overheads as a file needs to be read before the initial caching decision at fetch time can be made (rather than just reading from memory).

I think using two files could work. I don't see this introducing overhead if the in-memory cache is building itself in the background from the files. What I mean by that is that you could implement your cache controller entity so that:

* It's available for use instantly
* It starts scanning files in the background at its own pace, processes them and indexes them in memory (i.e. "low priority cache build queue")
* When requested to fetch a cached tile:
    * If memory indexing is complete: check memory directly, don't use files
    * If memory indexing is still in progress:
        * check the file system
        * if a pair of files is found, add the info to the memory cache currently building itself (i.e. "high priority cache build queue")
* Eventually, the background files scan completes and the cache is entirely stored in memory

With this approach, you might fetch cached files information from the disk when needing them, but that disk read was going to happen at some point anyways during the indexing.

If you have a cache of 10k files, it might mean that on cache initialization, you read the cache info from 10k files, which is really not ideal...

An alternative approach to this could be to lazy-access the cached files so you only read / discard them when attempting to fetch the cache entry. To keep track of the cache size without having to scan all files, you could have a unique file storing the total size value, and that would only be a few bytes.

## A few more thoughts about this

* HTTP headers returned by OSM dictate how long we want to cache something, is that it?
* I'm assuming we'll want to reduce the frequency at which we write the registry (I saw some de-bounce code in there, not sure if it was for that), if you have 10k cached tiles, are you going to write the entire 10k cache information to the disk every time you save your registry?

JaffaKetchup commented:
Just to be clear, you're suggesting removing the persistent registry? I'm happy with this in theory, but in practise, I think it could be a little more difficult.
The advantage is that storing metadata separately for each tile would allow instant starting of the caching mechanism, however, the suggested in-memory registry would be built much slower (have to read through many files).

I think having the single registry is better in this respect - reading 10k files (which would only cover 5k tiles) doesn't seem like the best option. If we reduce the duration of the size limiter (see below), we can bring the initialisation time down significantly.

We can also provide an optional 'preload' method to be called in `main` (for example).

> An alternative approach to this could be to lazy-access the cached files so you only read / discard them when attempting to fetch the cache entry.

I'm not quite sure what you mean by this exactly, could you explain a little more? If you mean we just check the metadata file for every tile load, that is an option, but it would make tile loading much slower.

> To keep track of the cache size without having to scan all files, you could have a unique file storing the total size value, and that would only be a few bytes.

This sounds like a good idea. The size limiter is by far the largest slow down in initialisation. But if we are over the limit, ordering the tiles by size and last modified is also really slow. Is there a better way to do this?


Is there any way we can get the best of both worlds? We don't need to be too careful with RAM particularly IMO. Is there a possibility of somehow using both a centralised registry and decentralised metadata file structure - or is that asking for too much (I feel like it might be)?

---

In terms of file names, why is a hash (or any other generation function) not good? It means no worrying about particular URLs or coordinates. I would propose the file names `<hash>` (we don't store an extension for ease of use) and `<hash>.metadata`.

> What do you mean by costly? More costly than fetching the data from the web with an HTTP request over the network? The tiles are fairly small and light in volume (~50kb for a 256x256 tile) - I can't imagine fetching cached tile for the entire screen of a device would exceed 10MB of reading on the disk, and that would almost always be faster than re-fetching them from the network anyways, right?

I was referring only to writing to the persistent registry. Opening it for each update would be unnecessary.

> What kind of overhead are you thinking about? I'm not super knowledgeable about these things so I'd be curious to have insights about this. I have a hunch that isolates might have noticeable value in intensive I/O operations, but given the size and number of the tiles we're dealing with here, I feel like it will not make a difference for the user, especially if you compare it with network-based requests. It might be worth keeping them for later if needed, instead of preemptively making them part of the architecture, but it's just a thought.

From my (admittedly non-profiled) experience when writing FMTC, putting I/O in a separate thread definitely reduced jank.

Also https://dart.dev/tools/linter-rules/avoid_slow_async_io exists - although I will admit I thought it applied to all synchronous I/O methods, not just the listed ones. See also https://github.com/dart-lang/sdk/issues/19156.

One thing that's more difficult when dealing with this kind of traffic is that the tile updates are relatively small each (but not totally insignificant), but there's a lot of them and they come in very quick bursts. If you've got an open HTTP connection, making a bunch of tile requests to fill the screen will result in a burst of responses.

> HTTP headers returned by OSM dictate how long we want to cache something, is that it?

Yes, or by any other server. The OSM ones in particular seem a little strange (serving tiles repeatedly that have already expired), but that's an issue at the source.

> I'm assuming we'll want to reduce the frequency at which we write the registry (I saw some de-bounce code in there, not sure if it was for that), if you have 10k cached tiles, are you going to write the entire 10k cache information to the disk every time you save your registry?

If we keep using a registry, then yeah, we don't want to update it for each individual tile (especially if we're using JSON as we need to write and flush the whole file each time). My initial idea was to write it and queue up any updates that occur whilst writing, then write again, but I couldn't get this working well, so I just resorted to a 50ms debounce.

androidseb commented:
> Just to be clear, you're suggesting removing the persistent registry?

Yes, or at least I'm hoping we can achieve this somehow. But maybe there is no good way around it...

---

> > An alternative approach to this could be to lazy-access the cached files so you only read / discard them when attempting to fetch the cache entry.
> 
> I'm not quite sure what you mean by this exactly, could you explain a little more? If you mean we just check the metadata file for every tile load, that is an option, but it would make tile loading much slower.

Yes that's what I mean. I understand this is "much slower" than direct RAM access, but we're still talking about sub-16-milliseconds frame time to even be noticeable (even in the bursts you mentioned?) by the user hopefully? I/O being "slow" is not a problem if the user can't notice. It isn't "slow" as in "consuming more CPU + battery" slow, it's slow as in "extra I/O idle delays" slow.

---

> > To keep track of the cache size without having to scan all files, you could have a unique file storing the total size value, and that would only be a few bytes.
> 
> This sounds like a good idea. The size limiter is by far the largest slow down in initialisation. But if we are over the limit, ordering the tiles by size and last modified is also really slow. Is there a better way to do this?
> 
> Is there any way we can get the best of both worlds? We don't need to be too careful with RAM particularly IMO. Is there a possibility of somehow using both a centralised registry and decentralised metadata file structure - or is that asking for too much (I feel like it might be)?

Yeah getting the best of both worlds would be good :smile:. I think the top priority here should be avoid sending extra HTTP requests when we have the cache. The long initialization time is an incentive to skip the cache at the beginning, but that's unfortunately the moment where the user gets the most tiles (the entire screen), and at scale, this implementation would mean we're effectively not minimizing (only reducing) the out of the box impact of flutter_map on the OSM servers.

To be clear, my suggestion of the direct file system approach was primarily driven by the intent to get around the initialization time, because I feel like we should not skip the cache, and I would like to find a solution where the user doesn't have to wait several seconds before seeing tiles.

At the very least, the hybrid approach could have the long initialization time, but during that time, instead of using direct network requests as a backup, we could use direct file system requests as a backup, because we know where the tile cached file will be located?

---

> In terms of file names, why is a hash (or any other generation function) not good? It means no worrying about particular URLs or coordinates. I would propose the file names <hash> (we don't store an extension for ease of use) and <hash>.metadata.

That's fine, I think hashes for file names would work too. I thought having a file name structure where you can trace back which tile a specific file is linked to just from its name could be useful, but I can't come up with any concrete use case, so it's probably not useful at the moment.

---

> From my (admittedly non-profiled) experience when writing FMTC, putting I/O in a separate thread definitely reduced jank.

Interesting, I wonder if this is related to the debugging setup specifically.

---

> Also https://dart.dev/tools/linter-rules/avoid_slow_async_io exists - although I will admit I thought it applied to all synchronous I/O methods, not just the listed ones. See also https://github.com/dart-lang/sdk/issues/19156.

I think it depends what you mean by "slow". I may be mistaken, but the way I understand it, there are two kinds of "slow" we're looking in this context:

1. CPU load slowness: deriving a cryptographic key many times is slow because it requires a lot of calculations - running long sync operations using CPU on the main thread will cause jank
2. I/O wait slowness: reading from a file is slow, because you need to wait for I/O - running these operations with await / async will not cause jank, because the operations in question are already being sent to another thread under the hood (which is why I'm saying we may not need isolates, because those are meant for CPU heavy operations)

For example, the link you shared mentions "Reading files asynchronously is 40% slower", but that's the total round-trip time for a single operation, because that operation is waiting a lot for things to come back between threads. If you were running 10 of those operations at the same time, they would probably not take 10x the time to execute, probably closer to the same time as a single operation. Because most of the time needed to perform the operation is waiting between threads.

In our specific setup with cache, the file operations I'm talking about are all affected by "I/O wait slowness", so fetching a single tile's data might take 150ms, but completing a quick burst of 100 of these operations running in parallel might not take much longer than 200ms to complete... And I'm using 150ms as an example, but it's probably less, and most certainly faster than a network-based request.

Don't take my word for it, we should  look into it, but I'm just saying it could be worth considering.

---

> If we keep using a registry, then yeah, we don't want to update it for each individual tile (especially if we're using JSON as we need to write and flush the whole file each time). My initial idea was to write it and queue up any updates that occur whilst writing, then write again, but I couldn't get this working well, so I just resorted to a 50ms debounce.

For the registry serialization, there might be good libraries to handle updates more efficiently, I've used sqflite in my project, but I feel like adding an SQL dependency to flutter_map for this is overkill. Maybe [hive](https://pub.dev/packages/hive) could be an interesting choice, even if it's to look at the code to find out how it's implemented.

---

One more note: it looks like you're abstracting the cache implementation so we can write our own implementation of it, I think that's good, because it will make contributions to alternate cache-focused plugins (those that need extra dependencies) easier down the line.

JaffaKetchup commented:
@androidseb So just because I'm not certain what we're suggesting exactly, we're suggesting using both a registry and an individual metadata file for each tile? I know that's perhaps what I was suggesting, but I've had a think about it, and that sounds really difficult to manage.

This would be my new suggestion:
 * Revert to waiting for the cache to be ready before tiles can be loaded
 * Recommend to users to call a method in `main` to preload the cache if performance matters
 * Remove the waiting for the size limiter - handle this in the background somehow (this is likely to be quite complex, but I think I can make it work reasonably well?)
 * Store the total cache size in another file
 * Potentially reduce the isolate work - I would want to keep the isolates for the persistent registry writer and total cache size writer at least

How does that sound to you (and @fleaflet/maintainers)? Basically, if the user doesn't call to preload, the delay is just the time taken to parse the registry.

atototenten commented:
why not use lmdb(3x faster than sqlite ,and 10x faster than file-systems) instead of file-system ?

lmdb is generally :
 - compared to sqlite :
  - 3-5 times faster
  - 2-3 times more space-efficent
 - compared to file-system
  - 10-15 times faster
  - 5 times more space-efficient

thanks

JaffaKetchup commented:
Yeah I have some experience with lmdb through Isar. For what I was trying to do with it, I had too many issues to put it in the core - maybe it would be more stable for this kind of use case.

But there's also the consideration to be made to keep the package lightweight - chances are, a production app might already use a library which depends on sqflite.

JaffaKetchup commented:
That does look like an encouraging package, but I would want more existing usage for it before adding it as a transitive dependency to all our users.

atototenten commented:
why not internalize/fork it ?

or maybe port it it dart ,using google-gemini then review ,because c is a lot like dart ,so it should be smooth ,hence it could become 10 - 15 kB ,compared to sqlite's 4 MB

atototenten commented:
why not internalize/fork it ?

or maybe port it to dart ,using google-gemini then review ,because c is a lot like dart ,so it should be smooth ,hence it could become 10 - 15 kB ,compared to sqlite's 4 MB

JaffaKetchup commented:
> why not internalize/fork it ?

That is infeasible. Our maintenance resources are stretched as it is, without having to manage an entirely new database project with no real low-level database experience.

> or maybe port it to dart ,using google-gemini then review ,because c is a lot like dart ,so it should be smooth ,hence it could become 10 - 15 kB ,compared to sqlite's 4 MB

We cannot just use a generative AI agent to perform code review. I personally stay away from generative AI for development, but even besides that, we cannot entrust the stability of such a large and complex project to AI.

---

@androidseb Looking at sqflite, we would need to use the layer to add compatibility for more platforms which uses sqlite3. So we may as well use sqlite3. Looking at that, I'm very unsure about adding sqlite3_flutter_libs, as this will put the package size way up (includes its own sqlite binaries), and making it work using the OS provided sqlite seems like a good way to introduce difficult to debug issues.

For the time being, I'm going to keep pursuing JSON. The size monitor has made a very large reduction to startup times, so that's good. I'm going to see if I can then make the size reducer run in the background without delaying the startup.

JaffaKetchup commented:
I've also changed the structure quite a bit in the latest commits. Now users can provide their own caching providers to `NetworkTileProvider` if they want. It should also make it easier to integrate into the cancellable TP.

JaffaKetchup commented:
Here's some very rough performance stats:

* To drop 350MB/31k tiles down to 1MB: ~2.5s size limiter, with a total init time ~3s 
* To load 11.2k tiles (registry 2MB) without limiting: ~470ms
* To load 9k tiles (registry 1.6MB) without limiting, but having to regenerate the size limiter: ~450ms

If we want to get the startup times down, I think the only option is to use a database at this point.

@androidseb @fleaflet/maintainers I think it's a good time for fresh feedback! I can try implementing a DB - but I'm a bit cautious about making it work seamlessly everywhere and keeping the weight down. What do you think about the new implementation style? Are there any other ways to get some more performance out? Thanks :)

atototenten commented:
LMDB is just 10k lines-of-code in c ,should be roughly 5-7 k l.o.c. in dart

maybe 10 kB in final binary

note that LMDB is more of storage-engine than a d.b. ,hence more suited to the project

flutter seems to be the future ,so i think this would also help the community

thanks

atototenten commented:
may i get some details ,regarding the requirements ?

would the data be add-only ,and never remove ,or simply re-create the data-file after a specific threshold is reached (of wastage) (used by both realm and sqlite) ?
then we can remove the paging and garbage-collection(of pages containing obsolete data) ,hence its at most 1-3 l.o.c. and have the impl. (outdated by 2 years) ,which can be refined .its as efficient as it can ,with minimum mgmt. portion ,actually a storage-framework to create d.b.s

androidseb commented on lib/src/layer/tile_layer/tile_provider/network/caching/tile_metadata.dart:
```diff
@@ -0,0 +1,79 @@
+        if (etag != null) 'etag': etag!,
+      };
+
+  @override
+  int get hashCode =>
+      Object.hash(lastModifiedLocally, staleAt, lastModified, etag);
+
+  @override
+  bool operator ==(Object other) =>
+      identical(this, other) ||
+      (other is CachedMapTileMetadata &&
+          lastModifiedLocally == other.lastModifiedLocally &&
+          staleAt == other.staleAt &&
+          lastModified == other.lastModified &&
+          etag == other.etag);
```
I personally find the [equatable](https://pub.dev/packages/equatable) package to handle these == and hashCode operators really good. This is just an idea in case you haven't heard about it, but it would essentially replace your hashCode and == functions with this:

```dart
class CachedMapTileMetadata extends Equatable {
//...
  @override
  List<Object?> get props => <dynamic>[
        lastModifiedLocally,
        staleAt,
        lastModified,
        etag,
      ];
```

This is just an idea in case it helps, feel free to resolve this comment

JaffaKetchup replied:
It might make sense to use equatable, but probably best to introduce it all at once in a separate PR. I've avoided it before just because I don't want to take the time to set it up everywhere (even though it's faster in the long term). I've looked for a good IDE plugin in the past, but I haven't found a reliable one.

androidseb commented:
> [...]
> I'm very unsure about adding sqlite3_flutter_libs, as this will put the package size way up
> [...]
> For the time being, I'm going to keep pursuing JSON
> [...]
> Now users can provide their own caching providers to NetworkTileProvider if they want.

Great, I think this makes perfect sense. Keeping the base package light while allowing for flexibility to implement more advanced and heavier caching packages makes perfect sense to me. As long as we can mitigate the impact on OSM servers while getting decent UX results that will not turn down library consumers, and have the straightforward and low-cost-to-maintain JSON approach... As I wrote earlier, I'd rather see you spend your time on valuable mapping features, we plugin developers can tackle these separate complicated advanced caching problems in separate plugin packages :-)

---

> [...] Here's some very rough performance stats [...] load 11.2k tiles (registry 2MB) without limiting: ~470ms

Are you saying there is about half a second delay on the startup time? This seems very acceptable if you compare this to fetching tiles from the network... Just wondering, did you test this in a production release on a real physical device? From my experience, the Android emulator can be deceivingly fast.

---

> I can try implementing a DB - but I'm a bit cautious about making it work seamlessly everywhere and keeping the weight down. What do you think about the new implementation style?

I agree, keeping it lightweight no DB is the way to go if we can have an acceptable tradeoff. A 470ms startup time seems very acceptable to me.

---

Apologies I don't have time to test things out (things are fairly busy for me at the moment), I've quickly glanced through the code, but at a high level the approach you're proposing makes sense to me. I hope these inputs help.

JaffaKetchup commented:
@atototenten 

> LMDB is just 10k lines-of-code in c ,should be roughly 5-7 k l.o.c. in dart

That is kinda surprising that it's that small, I would've expected it to be larger! I would be interested in a Dart port, but it's unfortunately not something that I would be comfortable doing outside of 'for fun' at the moment. Now that this functionality allows swapping of the caching provider, it would be easy to change in future.

> may i get some details ,regarding the requirements ?

Sure. This is a non-critical cache which stores lots of 10-50KB blobs very frequently, plus some metadata. So the requirements are to have very fast creates (or at least handle a burst well) and updates. Deletion is not important - it is currently implemented, but should never be used (it's used when the registry somehow gets out of sync, which shouldn't happen).
A cap on the size is useful, but other than that, considering what happens when data is removed is not necessary. If data is removed, all the data is removed. Reliability/robustness is not very important, so long as the cache can be reset.

I think LMDB could be something to explore in future, but it's not feasible at the moment. But maybe I have some free time at some point :D Thanks for the suggestions though, it's certainly useful to consider.

---

@androidseb

> Are you saying there is about half a second delay on the startup time?

I'm testing this on a Windows build. I think it's fair to say the longest time is the parsing of the JSON, so I think it would be fairly stable across devices. Also see https://github.com/dart-lang/sdk/issues/51596 & https://github.com/dart-lang/sdk/issues/51779.

This is also the delay before we make the first network requests. So it looks longer to the user. But I would be comfortable saying it would go unnoticed in all feasible registry sizes when awaited in the app startup, and maybe noticed but usually less than 1 sec if startup is performed just before the first tile load. Acceptable hopefully - maybe some stronger hints in documentation.

There is also a balance between the max size we impose (by default) and time spent 'size limiting'. We won't need to size limit (which is the expensive operation) for longer with a higher max size, but parse times will keep growing. It's current at 1GB - wdyt about changing this?

Thanks for all your feedback :)

androidseb commented:
Again, thanks for doing this work, this looks good and I'll look into leveraging these changes in a future app update.

I think it would be good to test worst case scenarios to get an idea, so what I have in mind for testing (when I get to it, you might do this way before me :smile:) is something like this:

* An app for which the map view is the main view, so the map is displayed on startup (e.g. I'll have this problem with my app) - this negates the effects of "trigger the parse ahead", because the app starts with the map view visible
* Running on an old, slow-ish Android device (a Windows PC is probably much faster)
* With the minimum reasonable cache size we'd like to see (e.g. 1GB)
* With the cache full (i.e. max parse time)

And then if I see the cache is fast enough, all good, otherwise I'll have to figure out ways to adjust the implementation, so I'll probably contribute back either core library improvements to your existing cache implementation, or if heavier dependencies are needed, improvements to an existing cache plugin, or the creation of a new one, but hopefully that won't even be needed.

JaffaKetchup commented:
I wonder if flatbufs or protobufs are a better option...

JaffaKetchup commented:
@androidseb I've changed the registry format from JSON to a FlatBuffer. I've also moved some stuff within the worker isolates around, and added a shortcut to check the size monitor size on initialisation without starting an isolate if it's not necessary to run the limiter.

Here's running in profile mode on my Windows again (which I note puts the `getApplicationCacheDirectory` method from 50ms to 0ms interestingly):

 * 30.2k tiles in 539MB, stored with a 3.6MB registry: total initialisation time ~300ms
 * 50.5k tiles in 812MB, stored with a 6MB registry: total initialisation time: ~500ms
   * Compared to approx the same time to load 11K tiles with a 2MB registry before
 * Initialisation time grows much slower than before
 * A good rule of thumb seems to be 100ms/10k tiles (at least for me)
 * A little more difficult to get a good benchmark on mobile - I still haven't done this

That's much better IMO. Even without moving the initialisation delay, I would say that it's barely noticeable until it reaches ~500ms. Maybe we drop our default limit down to 800MB?

JaffaKetchup commented:
I've exposed the tile count at initialisation from the API. In the example app I've added this count to the menu header, alongside the time it took to initialise.

If you have free time, it should now be relatively easy to test! Just download the APK from the actions output for the latest commit.

JaffaKetchup commented:
@androidseb Testing on my Android (https://www.gsmarena.com/xiaomi_redmi_note_13_pro-12581.php).

* Overall the init times are much more varied - I restarted the app a few times on each and took the extremes but without any real outliers.
* 15.5k tiles in 130ms - 220ms
* 17k in 150ms - 260ms
* 21k in ~230ms
* 24.6k in 220ms - 320ms

I haven't gone higher, but I feel comfortable saying that it's still relatively performant. Not quite 100ms:10k, but still close.

---

One thing I have found out is that FlatBuffers can be made to work with a binary search. So you wouldn't require a seperate in-memory registry - until you consider that you have to write in order every time, which does require a full sort of every single tile for every single update, or an in-memory registry. So it may be possible to implement it so that reads can happen using binary search whilst a registry is built in the background.

I had a start at implementing that, but it makes things much much more complex (especially since there is no way to validate the FlatBuffer, so any read at any time might not work as expected if there's a real corruption) and for probably an overall small gain in the end. Maybe I can look into it once I have more time if there's a real need.

I'm going to drop the cache limit down to 800 MB.

JaffaKetchup commented:
I'm just running some more tests comparing against JSON. I realised it may have not been a fair comparison before - I was not running in profile/release mode I don't think.

As it turns out, I've just loaded 50k tiles with a JSON registry consistently in 300ms on Windows. Yes, the registry size is definitely larger, but with more efficient encoding, I think we can get that down, and maybe the time too.

I may have made a mistake switching to flatbufs 😭 - more testing on Android to come tomorrow before I make a decision whether to move back or not. But it's strange that for both methods, the difference between debug, profile, and release is so large. Like easily 3-5x slower.

JaffaKetchup commented:
Back to JSON, getting a ratio of 40k:100ms on Windows and 10k:100ms on Android (with a slightly noticeably worse ratio at lower numbers (where the constant time of the isolate spawning has more of an overall effect). Debug mode is ~10x slower.

The size limiter on Windows does 680 MB/40.5k tiles down to 1 MB in a total time of a little under 4 seconds. Not really an easy way to make this faster, it's just a lot of I/O.

I also removed one thing we store which wasn't actually very useful very often, which has also brought the JSON size ratio down to a slightly better than 10k:1 MB.

I think I'm going to leave it there for now. I can't think of any way to get that down any more, given it's now clearly very device dependent.

JaffaKetchup commented:
https://github.com/dart-lang/http/pull/1773 👀

androidseb reviewed (commented):
@JaffaKetchup no worries, I'm happy if I can help - amazing news about the updated performance, 0 startup time sounds great, thanks for your work :100: 

I've taken a quick look at the code because I was curious to see how you achieved this, I added a couple of comments for consideration in case it helps, feel free to resolve them if they're not relevant.

JaffaKetchup commented on lib/src/layer/tile_layer/tile_provider/network/caching/built_in/impl/native/workers/tile_and_size_monitor_writer.dart:
```diff
@@ -0,0 +1,151 @@
+  final allocatedInt64Buffer = Uint8List(8);
+  final allocatedUint16Buffer = Uint8List(2);
+
+  await for (final val in receivePort) {
```
Done this, it does make it a little clearer with the early returns/continues and forces it to be all sync.

androidseb commented on lib/src/layer/tile_layer/tile_provider/network/caching/built_in/impl/native/workers/tile_and_size_monitor_writer.dart:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    // We write the new tile bytes to the file and truncate it to the end
+    ram.writeFromSync(tileBytes);
```
I agree, maybe the checksum is over the top, especially since we're writing bytes in a linear way here.

I think at least writing the "bytes size" information instead of the checksum, and then checking it back when reading might help: if you abort a PNG image write mid-way, you might end up with a "valid" decodable PNG image that is still kind of "corrupted" because it's missing bytes data, and your method of checking for decoding failure might not detect this kind of corruption.

From my testing so far, this does not seem to happen: I tested a cut off PNG image and it fails to decode in Flutter, but future upgrades to the Flutter engine might make it more resilient to these decoding errors, and then your reliance on "partial image decode failing" might become a problem.

By experience I can tell this use case of writes being interrupted mid-way happens in real life (user force-quitting the app, unrelated app crash, phone battery dying, etc.).

To illustrate, I created a 9KB 1920x1080 green background PNG image, then created a bytes-cut-off version of it at 7KB with a terminal bash command `split -b 7k test.png`, here is the result:

|full image|corrupted image|
|-|-|
|![test](https://github.com/user-attachments/assets/1e27af52-60a4-4a8c-9fee-c6a06b85509b)|![test-cutoff](https://github.com/user-attachments/assets/dc0e6d7b-6a35-4e2a-8535-02e583e3909f)|

You can see the browser is more flexible in its decoding and displays the partial image. I don't think it's a big deal because it will happen fairly rarely, and the Flutter decoding of images seems to fail on those cut-off images today.

I hope these insights help.

JaffaKetchup replied:
I wasn't aware that a partial PNG could be decoded! I'll update the format to store the tile size as a Uint32, just before the tile bytes. Then we can compare lengths.

JaffaKetchup commented on lib/src/layer/tile_layer/tile_provider/network/caching/built_in/impl/native/workers/tile_and_size_monitor_writer.dart:
```diff
@@ -0,0 +1,322 @@
+import 'package:path/path.dart' as p;
+
+/// Isolate worker which writes tile files, maintains the size monitor, and
+/// (if necessary) starts the size reducer
```
Poor habit of mine I know :D

JaffaKetchup commented on example/lib/pages/abort_unnecessary_requests.dart:
```diff
@@ -0,0 +1,71 @@
+                  urlTemplate: 'https://tile.openstreetmap.org/{z}/{x}/{y}.png',
+                  userAgentPackageName: 'dev.fleaflet.flutter_map.example',
+                  tileProvider: NetworkTileProvider(
+                    abortUnneededRequests: _abortingEnabled,
```
Done!

JaffaKetchup commented on lib/src/layer/tile_layer/tile_provider/network/tile_provider.dart:
```diff
@@ -0,0 +1,133 @@
+  final bool _isInternallyCreatedClient;
+
+  @override
+  bool get supportsCancelLoading => true;
```
IMO the plan is to leave this as is for now. When we release #2120, there's going to be bucket loads of breaking changes and renamings (including stuff in this PR most likely, like "MapCachingManager"). That PR does assume abortability throughout.

josxha replied:
Sounds good to me. 👍🏼I agree, there is no use to direct the user to a new API that is not yet included.

josxha commented on lib/src/layer/tile_layer/tile_provider/network/caching/built_in/impl/native/README.md:
```diff
@@ -0,0 +1,44 @@
+
+Whilst it is being calculated (which should happen on the first initialisation of the cache, or when required as above), writes must be delayed. Reads can still occur.
+
+Named 'sizeMonitor.bin'.
```
this is really a nit, decide your self if it's worth the time to change it. 😅

JaffaKetchup replied:
I did this intentionally:
 * It differentiates it from tiles a little
 * We do define the format of the size monitor ourselves; we don't define the tile formats
 * Tiles don't carry extensions because doing so would not be worth the effort

JaffaKetchup commented on lib/src/layer/tile_layer/tile_provider/network/image_provider/image_provider.dart:
```diff
@@ -0,0 +1,326 @@
+
+  @override
+  int get hashCode =>
+      Object.hashAll([url, if (fallbackUrl != null) fallbackUrl]);
```
Ermmmm...
This is something to definitely get right - I hate equality and hash codes :D I suppose that if it is `null`, then hashing `null` into every `url` would make no difference between providers. So I guess, yes, that would be OK. I guess I did it originally to more closely match the equality operator's implementation?

JaffaKetchup commented on lib/src/layer/tile_layer/tile_provider/network/caching/built_in/impl/native/README.md:
```diff
@@ -0,0 +1,44 @@
+# `BuiltInMapCachingProvider` storage spec
+
+The `BuiltInMapCachingProvider`, referred to as just 'built-in caching', is implemented using the filesystem for storage on native platforms.
```
Done!

JaffaKetchup commented on lib/src/layer/tile_layer/tile_provider/network/caching/built_in/impl/native/workers/size_reducer.dart:
```diff
@@ -0,0 +1,56 @@
+import 'package:meta/meta.dart';
+import 'package:path/path.dart' as p;
+
+typedef _SizeReducerTile = ({String path, int size, DateTime sortKey});
```
Done! Also squashed a bug that I hadn't noticed before in the size reducer.

josxha commented on lib/src/layer/tile_layer/tile_provider/network/caching/caching_provider.dart:
```diff
@@ -0,0 +1,51 @@
+  /// Tile providers should anticipate these exceptions and fallback to a
+  /// non-caching alternative, wherever possible repairing or replacing the tile
+  /// with a fresh & valid one.
+  Future<({Uint8List bytes, CachedMapTileMetadata metadata})?> getTile(
```
Consider to use a class instead of a record here (:

josxha reviewed (commented):
I not sure either if pub.dev allows a release with a pre release dependency. But we'll find out I guess. 😄

JaffaKetchup commented:
Trying to release gives a warning:

> * Packages dependent on a pre-release of another package should themselves be published as a pre-release version. If this package needs http version 1.5.0-beta, consider publishing the package as a pre-release instead.
  See https://dart.dev/tools/pub/publishing#publishing-prereleases For more information on pre-releases.

I'm happy to ignore this though. If it's a problem, we can release a new version without abortable support pretty quickly, and we allow users to turn it off (assuming I didn't mess up any non-abortable related stuff, but all tests are passing).

josxha commented on lib/src/layer/tile_layer/tile_provider/network/caching/built_in/impl/native/README.md:
```diff
@@ -0,0 +1,53 @@
+
+The format of the header is as follows:
+
+1. (position  0) 6-byte ASCII encoded string: file format identifier "FMBICT" ("FlutterMapBuiltInCacheTile")
```
Alright 👍
